File: dat-nuget/Saro.Dat/DatSignatureHmac.cs
using System.Security.Cryptography;

namespace Saro.Dat;

public class DatSignatureHmac : IDatSignature, IDisposable
{
    private readonly DatSignatureAlgorithm _algorithm;
    private readonly byte[] _key;
    private bool _disposed;

    private DatSignatureHmac(DatSignatureAlgorithm algorithm, byte[] key)
    {
        _algorithm = algorithm;
        _key = key;
    }

    public static IDatSignature FromKey(DatSignatureAlgorithm algorithm, byte[] key)
    {
        if (key is null)
            throw new DatException(DatErrorCode.ConfigArgumentInvalid, "signature key bytes are null");
        if (GetKeySize(algorithm) != key.Length)
        {
            throw new DatException(DatErrorCode.KeyInvalid,
                $"hmac key must be {GetKeySize(algorithm)} bytes for {algorithm.ToText()}, got {key.Length}");
        }
        return new DatSignatureHmac(algorithm, key);
    }

    public static IDatSignature Generate(DatSignatureAlgorithm algorithm)
    {
        byte[] key = new byte[GetKeySize(algorithm)];
        RandomNumberGenerator.Fill(key);
        return new DatSignatureHmac(algorithm, key);
    }

    private static int GetKeySize(DatSignatureAlgorithm algorithm) => algorithm switch
    {
        DatSignatureAlgorithm.HmacSha256Mfs => 32,
        DatSignatureAlgorithm.HmacSha384Mfs => 48,
        DatSignatureAlgorithm.HmacSha512Mfs => 64,
        _ => throw new DatException(DatErrorCode.ConfigAlgUnsupported, $"not an hmac signature algorithm: {algorithm}")
    };

    public DatSignatureAlgorithm Algorithm() => _algorithm;

    public byte[] Sign(byte[] body) => _disposed ? throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatSignatureHmac)) : _algorithm switch
    {
        DatSignatureAlgorithm.HmacSha256Mfs => HMACSHA256.HashData(_key, body),
        DatSignatureAlgorithm.HmacSha384Mfs => HMACSHA384.HashData(_key, body),
        DatSignatureAlgorithm.HmacSha512Mfs => HMACSHA512.HashData(_key, body),
        _ => throw new DatException(DatErrorCode.ConfigAlgUnsupported, $"not an hmac signature algorithm: {_algorithm}")
    };

    public bool Verify(byte[] body, byte[] signature)
    {
        if (_disposed) throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatSignatureHmac));
        try
        {
            byte[] computed = Sign(body);
            return CryptographicOperations.FixedTimeEquals(computed, signature);
        }
        catch (DatException)
        {
            throw;
        }
        catch (Exception e)
        {
            throw new DatException(DatErrorCode.SigBackend, "hmac verification failed to run", e);
        }
    }

    public byte[] ExportKey(bool verifyOnly = false)
    {
        if (_disposed) throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatSignatureHmac));
        if (verifyOnly)
        {
            throw new DatException(DatErrorCode.KeyVerifyOnlyUnsupported, _algorithm.ToText());
        }
        return (byte[])_key.Clone();
    }

    public bool Signable() => true;

    public bool SupportVerifyOnly() => false;

    public object Clone()
    {
        if (_disposed) throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatSignatureHmac));
        return new DatSignatureHmac(_algorithm, (byte[])_key.Clone());
    }

    IDatSignature IDatSignature.Clone() => (IDatSignature)Clone();

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        CryptographicOperations.ZeroMemory(_key);
    }
}

File: dat-maven/src/main/kotlin/me/saro/dat/signature/DatSignatureHmac.kt
package me.saro.dat.signature

import me.saro.dat.exception.DatErrorCode
import me.saro.dat.exception.DatException
import me.saro.dat.signature.DatSignatureAlgorithm.*
import org.bouncycastle.crypto.Digest
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.digests.SHA384Digest
import org.bouncycastle.crypto.digests.SHA512Digest
import org.bouncycastle.crypto.macs.HMac
import org.bouncycastle.crypto.params.KeyParameter
import java.security.MessageDigest
import java.security.SecureRandom


class DatSignatureHmac private constructor(
    private val algorithm: DatSignatureAlgorithm,
    private val keyParameter: KeyParameter,
    private val secret: ByteArray,
) : DatSignature {

    companion object {
        private val random = SecureRandom();

        internal fun getGeneralDigest(algorithm: DatSignatureAlgorithm): Digest {
            return when (algorithm) {
                HMAC_SHA256_MFS -> SHA256Digest()
                HMAC_SHA384_MFS -> SHA384Digest()
                HMAC_SHA512_MFS -> SHA512Digest()
                else -> throw DatException(DatErrorCode.CONFIG_ALG_UNSUPPORTED, "not an hmac signature algorithm: $algorithm")
            }
        }

        internal fun getKeySize(algorithm: DatSignatureAlgorithm): Int {
            return when (algorithm) {
                HMAC_SHA256_MFS -> 32
                HMAC_SHA384_MFS -> 48
                HMAC_SHA512_MFS -> 64
                else -> throw DatException(DatErrorCode.CONFIG_ALG_UNSUPPORTED, "not an hmac signature algorithm: $algorithm")
            }
        }

        @JvmStatic
        internal fun fromKey(algorithm: DatSignatureAlgorithm, key: ByteArray): DatSignature {
            if (getKeySize(algorithm) != key.size) {
                throw DatException(
                    DatErrorCode.KEY_INVALID,
                    "hmac key must be ${getKeySize(algorithm)} bytes, got ${key.size}",
                )
            }
            return DatSignatureHmac(algorithm, KeyParameter(key), key)
        }

        @JvmStatic
        internal fun generate(algorithm: DatSignatureAlgorithm): DatSignature {
            val keyBytes = ByteArray(getKeySize(algorithm))
                .apply { random.nextBytes(this) }
            return fromKey(algorithm, keyBytes)
        }
    }

    override fun algorithm(): DatSignatureAlgorithm {
        return algorithm
    }

    override fun exportKey(verifyOnly: Boolean): ByteArray {
        if (verifyOnly) {
            throw DatException(DatErrorCode.KEY_VERIFY_ONLY_UNSUPPORTED, algorithm.text)
        }
        return secret
    }

    override fun verify(body: ByteArray, signature: ByteArray): Boolean {
        val mine = try {
            sign(body)
        } catch (e: DatException) {
            throw e
        } catch (e: Exception) {
            throw DatException(DatErrorCode.SIG_BACKEND, "hmac verification failed to run", e)
        }
        return MessageDigest.isEqual(mine, signature)
    }

    override fun sign(body: ByteArray): ByteArray {
        hmac.get().run {
            reset()
            update(body, 0 , body.size)
            return ByteArray(macSize)
                .apply { doFinal(this, 0) }
        }
    }

    override fun signable(): Boolean {
        return true
    }

    override fun supportVerifyOnly(): Boolean {
        return false
    }

    override fun clone(): DatSignature {
        return fromKey(algorithm, exportKey())
    }

    private val hmac: ThreadLocal<HMac> = ThreadLocal.withInitial {
        HMac(getGeneralDigest(algorithm))
            .apply { init(keyParameter) }
    }
}